File: typescript/src/resources/economic.ts
/**
 * Official PIA SDK - Economic Intelligence API Resource
 */

import type { HttpTransport } from "../http/transport";
import type {
  EconomicCalendarResponse,
  GetCalendarOptions,
  GetMacroMapOptions,
  MacroMapResponse,
  RequestOptions,
} from "../types";

export class EconomicResource {
  constructor(private readonly transport: HttpTransport) {}

  /**
   * Retrieves global Macro Maps choropleth data (Inflation, Unemployment, GDP, Policy Rates, PMI)
   * with sovereign rankings, historical timeline, and delta changes.
   */
  public async getMacroMap(options?: GetMacroMapOptions): Promise<MacroMapResponse> {
    const params = new URLSearchParams();
    if (options?.indicator) params.set("indicator", options.indicator);
    if (options?.period) params.set("period", options.period);

    const query = params.toString() ? `?${params.toString()}` : "";
    try {
      return await this.transport.request<any>(
        `/api/v1/macro/map${query}`,
        "GET",
        undefined,
        options
      );
    } catch {
      return await this.transport.request<any>(
        `/api/v1/economic/map${query}`,
        "GET",
        undefined,
        options
      );
    }
  }

  /**
   * Fetches global macroeconomic calendar events (NFP, CPI, interest rates, GDP).
   */
  public async getCalendar(options?: GetCalendarOptions): Promise<EconomicCalendarResponse> {
    const params = new URLSearchParams();
    if (options?.impact) params.set("impact", options.impact);
    if (options?.limit) params.set("limit", String(options.limit));

    const query = params.toString() ? `?${params.toString()}` : "";
    const raw = await this.transport.request<any>(
      "/api/v1/economic/calendar" + query,
      "GET",
      undefined,
      options
    );
    const events = raw?.events || raw?.items || raw?.data || [];
    return { ...raw, events: Array.isArray(events) ? events : [], items: Array.isArray(events) ? events : [], total: raw?.total ?? (Array.isArray(events) ? events.length : 0) };
  }

  /**
   * Retrieves macroeconomic indicators and series metadata.
   */
  public async getIndicators(options?: RequestOptions): Promise<any> {
    return this.transport.request<any>(
      "/api/v1/economic/indicators",
      "GET",
      undefined,
      options
    );
  }

  /**
   * Retrieves list of available macroeconomic indicator categories.
   */
  public async getCategories(options?: RequestOptions): Promise<any> {
    return this.transport.request<any>(
      "/api/v1/economic/categories",
      "GET",
      undefined,
      options
    );
  }

  /**
   * Retrieves list of supported sovereign countries for economic indicators.
   */
  public async getCountries(options?: RequestOptions): Promise<any> {
    return this.transport.request<any>(
      "/api/v1/economic/countries",
      "GET",
      undefined,
      options
    );
  }
}

File: python/pia/resources/economic.py
"""Official PIA SDK - Economic Intelligence Resource."""

from __future__ import annotations

from typing import Any, Dict, Optional

from ..transport import AsyncTransport, SyncTransport


class EconomicResource:
    """Synchronous Economic API resource."""

    def __init__(self, transport: SyncTransport) -> None:
        self._transport = transport

    def get_calendar(
        self,
        *,
        impact: Optional[str] = None,
        limit: Optional[int] = None,
    ) -> Dict[str, Any]:
        """Fetches global macroeconomic calendar events (NFP, CPI, interest rates)."""
        params: Dict[str, Any] = {}
        if impact:
            params["impact"] = impact
        if limit is not None:
            params["limit"] = limit

        return self._transport.request("/api/v1/economic/calendar", method="GET", params=params)

    def get_indicators(self) -> Dict[str, Any]:
        """Retrieves macroeconomic indicators and series metadata."""
        return self._transport.request("/api/v1/economic/indicators", method="GET")


class AsyncEconomicResource:
    """Asynchronous Economic API resource."""

    def __init__(self, transport: AsyncTransport) -> None:
        self._transport = transport

    async def get_calendar(
        self,
        *,
        impact: Optional[str] = None,
        limit: Optional[int] = None,
    ) -> Dict[str, Any]:
        """Fetches global macroeconomic calendar events (NFP, CPI, interest rates)."""
        params: Dict[str, Any] = {}
        if impact:
            params["impact"] = impact
        if limit is not None:
            params["limit"] = limit

        return await self._transport.request("/api/v1/economic/calendar", method="GET", params=params)

    async def get_indicators(self) -> Dict[str, Any]:
        """Retrieves macroeconomic indicators and series metadata."""
        return await self._transport.request("/api/v1/economic/indicators", method="GET")
